File: BOJ/2042_구간 합 구하기.py
# https://www.acmicpc.net/problem/2042

import sys
input = sys.stdin.readline


class FTree:
    def __init__(self, a):
        self.n = len(a)
        self.bit = [0] * (self.n + 1)

        for i in range(1, self.n + 1):
            self.bit[i] += a[i - 1]
            if i + self.lsone(i) <= self.n:
                self.bit[i + self.lsone(i)] += self.bit[i]

    def lsone(self, s):
        return s & (-s)

    def add(self, idx, val):
        while idx <= self.n:
            self.bit[idx] += val
            idx += self.lsone(idx)

    def prefix_sum(self, idx):
        answer = 0
        while idx > 0:
            answer += self.bit[idx]
            idx -= self.lsone(idx)
        return answer

    def sum(self, left, right):
        return self.prefix_sum(right) - self.prefix_sum(left - 1)


N, M, K = map(int, input().split())
arr = [0] * N
for i in range(N):
    arr[i] = int(input())

ft = FTree(arr)

for _ in range(M + K):
    a, b, c = map(int, input().split())
    if a == 1:
        ft.add(b, c - arr[b-1])
        arr[b - 1] = c
    elif a == 2:
        print(ft.sum(b, c))

File: BOJ/2042_구간 합 구하기.cpp
// # https://www.acmicpc.net/problem/2042

#include <bits/stdc++.h>
using namespace std;
#define REP(i, n) for (int i = 0; i < n; i++)
#define REPN(i, n) for (int i = 1; i <= n; i++)

typedef long long ll;
typedef vector<ll> vll;

#define LSOne(x) ((x) & -(x))
vll ft;

void build(int m, vll &a)
{
    ft.assign(m + 1, 0);
    REPN(i, m)
    {
        ft[i] += a[i];
        if (i + LSOne(i) <= m)
            ft[i + LSOne(i)] += ft[i];
    }
}

ll rsq(int i)
{
    ll sum = 0;
    for (; i; i -= LSOne(i))
        sum += ft[i];
    return sum;
}

ll rsq(int i, int j)
{
    return rsq(j) - rsq(i - 1);
}

void update(int i, ll v)
{
    for (; i < (int)ft.size(); i += LSOne(i))
        ft[i] += v;
}

int main()
{
    int n, m, k, x;
    int a, b;
    ll c;
    vll arr;

    scanf("%d %d %d", &n, &m, &k);
    arr.resize(n + 1);

    REPN(i, n)
    {
        scanf("%d", &arr[i]);
    }
    build(n, arr);

    REP(i, m + k)
    {
        scanf("%d %d %lld", &a, &b, &c);
        if (a == 1)
        {
            update(b, c - arr[b]);
            arr[b] = c;
        }
        else
        {
            printf("%lld\n", rsq(b, c));
        }
    }

    return 0;
}